interpn/src/multilinear/rectilinear.rs

```diff
@@ -46,44 +21,104 @@
 //! // Storage for output
 //! let mut out = [0.0; 2];
 //!
-//! // Do interpolation
-//! rectilinear::interpn(grids, &z, &obs, &mut out).unwrap();
+//! // Do interpolation, allocating for the output for convenience
+//! rectilinear::interpn_alloc(grids, &z, &obs).unwrap();
 //! ```
 //!
 //! References
 //! * https://en.wikipedia.org/wiki/Bilinear_interpolation#Weighted_mean
 use num_traits::Float;
 
-/// An arbitrary-dimensional multilinear interpolator / extrapolator on a rectilinear grid.
+/// Evaluate multicubic interpolation on a regular grid in up to 8 dimensions.
+/// Assumes C-style ordering of vals (z(x0, y0), z(x0, y1), ..., z(x0, yn), z(x1, y0), ...).
+///
+/// This is a convenience function; best performance will be achieved by using the exact right
+/// number for the MAXDIMS parameter, as this will slightly reduce compute and storage overhead,
+/// and the underlying method can be extended to more than this function's limit of 8 dimensions.
+/// The limit of 8 dimensions was chosen for no more specific reason than to reduce unit test times.
 ///
-/// Unlike `RegularGridInterpolator`, this method does not handle
-/// grids with negative step size; all grids must be monotonically increasing.
+/// While this method initializes the interpolator struct on every call, the overhead of doing this
+/// is minimal even when using it to evaluate one observation point at a time.
+#[inline(always)]
+pub fn interpn<T: Float>(
+    grids: &[&[T]],
+    vals: &[T],
+    obs: &[&[T]],
+    out: &mut [T],
+) -> Result<(), &'static str> {
+    MultilinearRectilinear::<'_, T, 8>::new(grids, vals)?.interp(obs, out)?;
+    Ok(())
+}
+
+/// Evaluate interpolant, allocating a new Vec for the output.
+///
+/// For best results, use the `interpn` function with preallocated output;
+/// allocation has a significant performance cost, and should be used sparingly.
+#[cfg(feature = "std")]
+pub fn interpn_alloc<T: Float>(
+    grids: &[&[T]],
+    vals: &[T],
+    obs: &[&[T]],
+) -> Result<Vec<T>, &'static str> {
+    let mut out = vec![T::zero(); obs[0].len()];
+    interpn(grids, vals, obs, &mut out)?;
+    Ok(out)
+}
+
+/// Check whether a list of observation points are inside the grid within some absolute tolerance.
+/// Assumes the grid is valid for the rectilinear interpolator (monotonically increasing).
+///
+/// Output slice entry `i` is set to `false` if no points on that dimension are out of bounds,
+/// and set to `true` if there is a bounds violation on that axis.
+///
+/// # Errors
+/// * If the dimensionality of the grid does not match the dimensionality of the observation points
+/// * If the output slice length does not match the dimensionality of the grid
+#[inline(always)]
```

Comment:
See if there's any perf difference between `inline` and `inline(always)`?

The former is much less risky, and usually there's not much to be gained by forcing it.

Reply:
When I added the `always`, it was a 30% speedup to some cases in benchmarks. That said, I'm starting to take the benchmarks with a bit more of a grain of salt